[sdk] Delegates alias computation to engine

Zaid-Ajaj opened the pull request:
Fixes #14 

Removes `AllAliases` function and instead constructs a`List<Pulumirpc.Alias>` in `PrepareAliases` (linear time) which populates the `RegisterResourceRequest` accordingly:
```csharp
if (prepareResult.SupportsAliasSpec) 
{
    request.Aliases.AddRange(prepareResult.Aliases);
}
else 
{
    var aliasUrns = prepareResult.Aliases.Select(a => a.Urn);
    request.AliasURNs.AddRange(aliasUrns);
}
```
Wasn't exactly sure how to test this other than constructing a lot (=1000) of aliases in `AliasesMocks` and _observe_ that it doesn't just hang for a while. 

All integration tests are succeeding locally except for `TestProvider` but that seems unrelated. Let's see what CI has to say about it.

Frassle commented:
Run `dotnet format` :) Also we should split that to it's own job so the unit tests can still run and give signal even if formatting is wrong.

Frassle commented on sdk/Pulumi/Deployment/Deployment_Prepare.cs:
```diff
@@ -146,6 +136,183 @@ void LogExcessive(string message)
+                    var aliasStack = await Resolve(resolvedAlias.Stack, Instance.StackName);
+                    var aliasProject = await Resolve(resolvedAlias.Project, Instance.ProjectName);
+
+                    if (string.IsNullOrEmpty(aliasName) || string.IsNullOrEmpty(aliasType))
```
This doesn't seem right. Quite often you'll give an alias with just a name and so type will be ""

Zaid-Ajaj replied:
In `CollapseAliasToUrn` we have
```csharp
var name = a.Name ?? defaultName;
var type = a.Type ?? defaultType;

// later... 

if (name == null)
    throw new ArgumentNullException("No valid 'Name' passed in for alias.");
if (type == null)
    throw new ArgumentNullException("No valid 'type' passed in for alias.");
```
Which suggested that both `name` _and_ `type` must be non-null so I wasn't sure. I will change it to:
```csharp
var aliasSpec = new Pulumirpc.Alias.Types.Spec
{
    Name = await Resolve(resolvedAlias.Name, resource.GetResourceName()),
    Type = await Resolve(resolvedAlias.Type, resource.GetResourceType()),
    Stack = await Resolve(resolvedAlias.Stack, Instance.StackName),
    Project = await Resolve(resolvedAlias.Project, Instance.ProjectName),
};
```

Frassle replied:
I'm pretty sure the engine is happy to take an Alias object with type or name as "".

Zaid-Ajaj commented:
@Frassle 
 - Using `dotnet run format-sdk` changes every file I have, removing a whitespace at the end 😓 
 - Integration test `TestProvider` is failing on the `destroy` step.

Frassle commented:
> Using dotnet run format-sdk changes every file I have, removing a whitespace at the end 😓

Odd? Shouldn't this be consistent across machines?

> Integration test TestProvider is failing on the destroy step.

OK leave the test off the test list for now, I'll get that fixed.

Zaid-Ajaj commented:
> Odd? Shouldn't this be consistent across machines?

@Frassle I've added a minimal .editorconfig which is now formatting the same locally and on CI

> OK leave the test off the test list for now, I'll get that fixed.

Removed that test the time being

> I'm pretty sure the engine is happy to take an Alias object with type or name as "".

Alright, then using empty string as default, see last commit:
```csharp
var aliasSpec = new Pulumirpc.Alias.Types.Spec
{
    Name = await Resolve(resolvedAlias.Name, ""),
    Type = await Resolve(resolvedAlias.Type, ""),
    Stack = await Resolve(resolvedAlias.Stack, Instance.StackName),
    Project = await Resolve(resolvedAlias.Project, Instance.ProjectName),
};
```

Frassle commented on .editorconfig:
```diff
@@ -0,0 +1,3 @@
+# C# files
```
Can you change this in it's own PR, it's pretty much ruined the diff for this PR.

Zaid-Ajaj replied:
OK let me try, because before adding the file I also got tons of diffs when I ran `dotnet run format-sdk`

Zaid-Ajaj replied:
That did not help at all 😞 now more diffs. Maybe `dotnet format` is not the same across machines _by default_ and it does need config?

Frassle replied:
Yes do the format change in another PR, I'll approve that then rebase this

Frassle commented on sdk/Pulumi/Deployment/Deployment_Prepare.cs:
```diff
@@ -146,6 +136,173 @@ void LogExcessive(string message)
+                    {
+                        Name = await Resolve(resolvedAlias.Name, ""),
+                        Type = await Resolve(resolvedAlias.Type, ""),
+                        Stack = await Resolve(resolvedAlias.Stack, Instance.StackName),
```
These can actually default to "" as well, the engine will fill them in.

Zaid-Ajaj replied:
done ✅
